bus-mapping/src/circuit_input_builder.rs

```diff
@@ -246,17 +309,39 @@ impl<'a, C: CircuitsParams> CircuitInputBuilder<C> {
         )?;
         tx.steps_mut().push(begin_tx_step);
 
-        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
+        let mut trace = geth_trace.struct_logs.iter().enumerate().peekable();
+        while let Some((i, geth_step)) = trace.next() {
```

Comment:
This design is a bit costly because of 
- `let mut cib = self.clone();`: clone circuit builder for every step
- `compute_param`: recompute everything while we only need max_rws()

Furthermore, there is one edge case unresolved by current logic: it will be possible that the max_rws are explode between `end_tx_step` and next `begin_tx_step` because current peekable design doenst check when step are located between end_tx and  next begin_tx.


I think we can avoid check max_rws every step by introducing `reserved_padding_rws` and we can set it a value like 512 or 1024 satisfying 2 criteria
- reserved_padding_rws `>` max rws contributing per evm steps.
- reserved_padding_rws `>` (end_tx + begin_tx rws) or (end_tx + end_block rws) or (end_chunk rws)

And we make psuedo code like this

``` rust
        assert!(self.chunk_ctx.rwc < circuits_params.max_rws() -  reserved_padding_rws)
        // Generate BeginTx step
        let begin_tx_step = gen_associated_steps(
            &mut self.state_ref(&mut tx, &mut tx_ctx),
            ExecState::BeginTx,
        )?;
        tx.steps_mut().push(begin_tx_step);
        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
            // check before process current step
            if self.chunk_ctx.rwc >= circuits_params.max_rws() -  reserved_padding_rws { 
                    // pre-process current step to get next ops
                    let mut cib_dryrun = self.clone(); 
                    let ops = gen_associated_ops(
                        &next_step.op,
                        &mut cib.state_ref(&mut cib_dryrun, &mut tx_ctx.clone()),
                        &geth_trace.struct_logs[*i..],
                    )?;
                   
                    let next_step = ops[0];
                    // set next step pre-state as end_chunk state
                    self.set_end_chunk(next_step);
                    self.commit_chunk(true);
                    self.set_begin_chunk(next_step);
            }
            // process current step
            let exec_steps = gen_associated_ops(
                &geth_step.op,
                &mut state_ref,
                &geth_trace.struct_logs[index..],
            )?;
            tx.steps_mut().extend(exec_steps);
        }
    ...
```

With that, we can also avoid hit max_rws between [end_tx, next_begin_tx], besides we should also avoid end_chunk/begin_chunk happend between [end_tx, next_begin_tx]

Reply:
Sounds good! I just made some updates, feel like with the window it should be okay with chunking happening btw  [end_tx, next_begin_tx], or the code will get kind of dirty